File: apps/grove/game-builder/src/tasks.ts
import { REQUEST_ID_HEADER, Task, type TaskId, type TaskStatusUpdate } from '@grove/api-contract';
import type { Env } from './env.js';

/**
 * `settled` is the task as it now stands; `refused` is @grove/api declining to move it, which is
 * what a redelivery of work somebody already finished earns.
 */
export type TaskWritten =
    { outcome: 'settled'; task: Task } | { outcome: 'refused' } | { outcome: 'unavailable' };

/**
 * Where this service says what it did with a task.
 *
 * The only call this process makes to @grove/api, and the reason it holds no database credential:
 * the transitions are checked there, so a builder that comes back from the dead and tries to settle
 * a task another box already finished is refused rather than overwriting it.
 */
export interface Tasks {
    advance(task: TaskId, update: TaskStatusUpdate, requestId: string): Promise<TaskWritten>;
}

/** Settling a task is a row update behind a bearer, not a compile: it answers or it is down. */
const SETTLE_TIMEOUT_MS = 5_000;

export function httpTasks(env: Env): Tasks {
    return {
        advance: async (task, update, requestId) => {
            const written = await fetch(`${env.API_URL}/v1/tasks/${task}`, {
                method: 'PATCH',
                headers: {
                    'content-type': 'application/json',
                    authorization: `Bearer ${env.FLEET_SECRET}`,
                    [REQUEST_ID_HEADER]: requestId,
                },
                body: JSON.stringify(update),
                signal: AbortSignal.timeout(SETTLE_TIMEOUT_MS),
            }).catch(() => undefined);

            if (written === undefined) return { outcome: 'unavailable' };
            // Told outright that this attempt lost, rather than left to infer it from a status:
            // a builder that keeps retrying a settled task is one that never acknowledges it.
            if (written.status === 404 || written.status === 409) return { outcome: 'refused' };
            if (!written.ok) return { outcome: 'unavailable' };
            return { outcome: 'settled', task: Task.parse(await written.json()) };
        },
    };
}

File: apps/grove/asset-upload-service/src/tasks.rs
//! Where this service says what it did with a task.
//!
//! The one call it makes to `@grove/api`, and the reason it holds no database credential: the
//! transitions are checked there, so a worker that comes back from the dead and tries to settle a
//! task another process already finished is refused rather than overwriting it.

use std::time::Duration;

use anyhow::{Context, Result};
use serde::Serialize;

/// Settling a task is a row update behind a bearer, not a transcode: it answers or it is down.
const SETTLE_TIMEOUT: Duration = Duration::from_secs(5);

/// The correlation id the rest of the fleet carries, so one task is one line across three services.
const REQUEST_ID_HEADER: &str = "x-request-id";

/// What the status route took.
///
/// `Refused` is `@grove/api` declining to move the task — already settled, or gone — which is the
/// one failure a worker must acknowledge rather than retry.
#[derive(Debug, PartialEq, Eq)]
pub enum Settled {
    Written,
    Refused,
    Unavailable,
}

#[derive(Serialize)]
struct StatusUpdate<'a> {
    status: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    detail: Option<Detail<'a>>,
}

#[derive(Serialize)]
struct Detail<'a> {
    #[serde(skip_serializing_if = "Option::is_none")]
    message: Option<&'a str>,
}

/// The status route, as this service sees it.
#[async_trait::async_trait]
pub trait Tasks: Send + Sync {
    async fn advance(&self, task: &str, status: &str, message: Option<&str>) -> Settled;
}

/// `@grove/api` over HTTP, behind the bearer its gate compares.
pub struct HttpTasks {
    client: reqwest::Client,
    api_url: String,
    fleet_secret: String,
}

impl HttpTasks {
    pub fn new(api_url: String, fleet_secret: String) -> Result<Self> {
        Ok(Self {
            client: reqwest::Client::builder()
                .timeout(SETTLE_TIMEOUT)
                .build()
                .context("building the API client")?,
            api_url,
            fleet_secret,
        })
    }
}

#[async_trait::async_trait]
impl Tasks for HttpTasks {
    async fn advance(&self, task: &str, status: &str, message: Option<&str>) -> Settled {
        let body = StatusUpdate {
            status,
            detail: message.map(|message| Detail {
                message: Some(message),
            }),
        };

        let answered = self
            .client
            .patch(format!("{}/v1/tasks/{task}", self.api_url))
            .bearer_auth(&self.fleet_secret)
            .header(REQUEST_ID_HEADER, task)
            .json(&body)
            .send()
            .await;

        match answered {
            Err(err) => {
                tracing::warn!(task, error = ?err, "task could not be settled");
                Settled::Unavailable
            }
            // Told outright that this attempt lost rather than left to infer it: a worker that
            // keeps retrying a settled task is one that never acknowledges its message.
            Ok(response) if response.status() == 404 || response.status() == 409 => {
                Settled::Refused
            }
            Ok(response) if response.status().is_success() => Settled::Written,
            Ok(response) => {
                tracing::warn!(task, status = %response.status(), "task not settled");
                Settled::Unavailable
            }
        }
    }
}
